### PR TITLE
fix: 

### PR DESCRIPTION
**Pull request by bitsandfoxes**
Resolves https://github.com/getsentry/sentry-dotnet/issues/2591

I don't understand how it even gets there. The property is guarded by null-coalescing operators. But the SDK should not just throw regardless.

So now we have a getter that just creates the context in case it's `null`.

### CODE REVIEW

**Review by jamescrosswell — COMMENTED** [review]
@bitsandfoxes I'm not sure I get this. Other than adding logging (when creating the dynamic sampling context from the propagation context) what has changed here?

**Comment by bitsandfoxes** [comment]
The SDK is now no longer throwing exceptions. Instead of being opinionated regarding the setter being called only once we defensively handle the dynamic sampling context creation in the getter.

**Comment by jamescrosswell** [comment]
> The SDK is now no longer throwing exceptions. Instead of being opinionated regarding the setter being called only once we defensively handle the dynamic sampling context creation in the getter.

Ah I see, there's no setter for DynamicSamplingContext anymore...

File: src/Sentry/SentryPropagationContext.cs [hunk 1]
```diff
@@ -33,13 +29,19 @@ internal class SentryPropagationContext
         TraceId = traceId;
         SpanId = SpanId.Create();
         ParentSpanId = parentSpanId;
-        DynamicSamplingContext = dynamicSamplingContext;
+        _dynamicSamplingContext = dynamicSamplingContext;
     }
 
-    public SentryPropagationContext()
+    public SentryPropagationContext() : this(null)
+    { }
```
Comment (jamescrosswell) [inline src/Sentry/SentryPropagationContext.cs:36]: When this constructor gets called in the `CreateFromHeaders` method, won't the TraceId and SpanId be empty? Is there ever a scenario where we'd want that to be the case? 

Even if we don't get it from upstream, I think we always want a TraceId to link this span to downstream events right?

File: test/Sentry.Tests/SentryPropagationContextTests.cs [hunk 2]
```diff
@@ -2,6 +2,43 @@ namespace Sentry.Tests;
...
+        var propagationContext = new SentryPropagationContext();
+        var firstDynamicSamplingContext = propagationContext.GetOrCreateDynamicSamplingContext(options);
+
+        var secondDynamicSamplingContext = propagationContext.GetOrCreateDynamicSamplingContext(options);
+
+        Assert.Same(firstDynamicSamplingContext, secondDynamicSamplingContext);
+    }
+
     [Fact]
     public void CreateFromHeaders_TraceHeaderNullButBaggageExists_CreatesPropagationContextWithoutDynamicSamplingContext()
```
Comment (jamescrosswell) [inline test/Sentry.Tests/SentryPropagationContextTests.cs:43]: I think we need another test where the traceheader is null and there is no baggage, to make explicit what our expectations are in that scenario.

File: test/Sentry.Tests/ScopeTests.cs [hunk 3]
```diff
@@ -98,7 +98,8 @@ public void Clone_NewScope_IncludesPropagationContext()
 
         var clone = sut.Clone();
 
-        Assert.Same(propagationContext, clone.PropagationContext);
+        Assert.Equal(propagationContext.TraceId, clone.PropagationContext.TraceId);
+        Assert.Equal(propagationContext.SpanId, clone.PropagationContext.SpanId);
```
Comment (jamescrosswell) [inline test/Sentry.Tests/ScopeTests.cs:102]: Maybe also assert that the references are not eqal (i.e. this is a clone... not the same object)?

**Review by jamescrosswell — CHANGES_REQUESTED** [review]
Looks good generally. I just had some questions about the SentryPropagationContext constructor but otherwise good to go I think.

**Review by jamescrosswell — APPROVED** [review]
The code itself looks good... once our CI build environment is healthy again 👍🏻
